File: tomosynthesis/submissions/accepted/tomosynthesis_ulf.cpp
//Solution by Ulf Lundström
#include <iostream>
#include <cstdio>
#include <cmath>
#include <vector>
#include <algorithm>
using namespace std;

typedef pair<double,bool> pdb;

int main() {
	int N;
	cin >> N;
	vector<double> x(N), y(N), r(N);
	for (int i = 0; i < N; ++i) {
		cin >> x[i] >> y[i] >> r[i];
	}
	vector<pdb> events;
	for (int i = 0; i < N; ++i) {
		for (int j = i+1; j < N; ++j) {
			double dx = x[i]-x[j], dy = y[i]-y[j];
			double a = atan2(dy,dx);
			if (a < 0) a += M_PI;
			double b = asin((r[i]+r[j])/sqrt(dx*dx + dy*dy));
			double start = a-b, end = a+b;
			events.push_back(pdb(start,true));
			events.push_back(pdb(end,false));
			if (start < 0) {
				events.push_back(pdb(start+M_PI,true));
				events.push_back(pdb(end+M_PI,false));
			}
			if (end > M_PI) {
				events.push_back(pdb(start-M_PI,true));
				events.push_back(pdb(end-M_PI,false));
			}
		}
	}
	double best = 0;
	int depth = 0;
	sort(events.begin(),events.end());
	events.push_back(pdb(events[0].first+M_PI,true));
	for (int i = 0; i < events.size()-1; ++i) {
		if (events[i].second)
			depth++;
		else {
			depth--;
			if (depth == 0)
				best = max(best,events[i+1].first-events[i].first);
		}
	}
	printf("%.9lf\n",best);
}

File: tomosynthesis/submissions/accepted/tomosynthesis_ulf.py
#!/usr/bin/env python3
from sys import stdin
from math import *

N = int(stdin.readline())
x = [0 for a in range(N)]
y = [0 for a in range(N)]
r = [0 for a in range(N)]
for i in range(N):
	x[i], y[i], r[i] = map(float,stdin.readline().split())
events = []
for i in range(len(x)):
	for j in range(i):
		dx = x[i]-x[j];
		dy = y[i]-y[j];
		a = atan2(dy,dx);
		if a < 0:
			a += pi
		b = asin((r[i]+r[j])/sqrt(dx**2+dy**2))
		start = a-b
		end = a+b
		events.append((start,True))
		events.append((end,False))
		if (start < 0):
			events.append((start+pi,True))
			events.append((end+pi,False))
		if (end > pi):
			events.append((start-pi,True))
			events.append((end-pi,False))
best = 0
depth = 0
events.sort()
events.append((events[0][0]+pi,True))
for i in range(len(events)):
	if events[i][1]:
		depth += 1
	else:
		depth -= 1
		if depth == 0:
			best = max(best,events[i+1][0]-events[i][0])
print('{:.9f}'.format(best))
